### PR TITLE
Burn YAML with fire (kind of)

### LINKED ISSUE #1004: Consider not handling YAML in APIserver
The APIserver processes JSON and YAML.
A vulnerability in the apiserver would be critical.

Compared to JSON:
- YAML has a crazy complicated SPEC.
- YAML has a go implementation that is more complex 
  - 14KLOC JSON vs 30KLOC YAML
- YAML has fewer tests in the go implementation
  - 8KLOC JSON vs 2.5KLOC YAML
- Has a track record of vulnerabilities 
   - YAML has fewer vulnerabilities, but this may be due to its less widespread use in critical software
   - 24 CVEs match YAML vs 66 JSON, using http://web.nvd.nist.gov/view/vuln/search

Therefore we should consider one of the following changes:
 1. Do not process YAML in the apiserver itself:
  1. Convert YAML to JSON in clients 
  1. Recommend that users needing to use YAML via `curl`, etc, can pipe YAML through simple yaml to json converter tool.
 1. Convert YAML to JSON in a less privileged proxy process (which does not have access to etcd).

### PR DESCRIPTION
**Pull request by smarterclayton**
Remove automatic decoding of YAML via core Codec and replace with opt-in decoding.

Partially addresses #1004

Needs an e2e run in case there's a missed spot.

### CODE REVIEW

File: pkg/runtime/codec.go [hunk 1]
```diff
@@ -16,11 +16,47 @@ limitations under the License.
...
 }
 
+// yamlCodec converts YAML passed to the Decoder methods to JSON.
+type yamlCodec struct {
+	// a Codec for JSON
+	Codec
+}
+
+// yamlCodec implements Codec
+var _ Codec = yamlCodec{}
```
Comment (jlowdermilk) [inline pkg/runtime/codec.go]: What's the point of this line?

**Comment by jlowdermilk** [comment]
lgtm, pending e2e run (in progress).

File: pkg/runtime/codec.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/runtime/codec.go]: To prove that yamlCodec implements Codec.  We started doing that in other places where there's a type that *should* implement a named interface.

----- Original Message -----

**Comment by bgrant0607** [comment]
Great, thanks.

FYI/FWIW, Kubelet already explicitly calls yaml.Unmarshal:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubelet/config/file.go#L161

**Comment by bgrant0607** [comment]
AFAICT, we're still not setting Content-Type in the client, and we're not documenting the accepted Content-Type in apiserver:
https://github.com/GoogleCloudPlatform/kubernetes/blob/32694a770ceb43786ab0a5284ad366643381c2af/pkg/apiserver/api_installer.go#L81

@nikhiljindal

**Comment by smarterclayton** [comment]
rebased

**Comment by bgrant0607** [comment]
Did we get results of the e2e run?

**Comment by smarterclayton** [comment]
Doing an e2e run now

**Comment by jlowdermilk** [comment]
e2e passing this morning (locally). This lgtm to merge.

**Comment by smarterclayton** [comment]
Merge away - don't wait for my tests (saves me a few hours)

**Comment by thockin** [comment]
This seems to have broken tests for pkg/kubelet/config

**Comment by smarterclayton** [comment]
Fixing that

**Comment by smarterclayton** [comment]
Crossed merges

**Comment by thockin** [comment]
Also seems to have broken integration

**Comment by smarterclayton** [comment]
Integration is fixed in https://github.com/GoogleCloudPlatform/kubernetes/pull/5775

----- Original Message -----

**Pull request merged**